Adding Locust as performance OQS tool

davidgca opened the pull request:
### Purpose:
The primary goal of this contribution is to enable Locust to perform quantum-safe load testing scenarios, utilizing quantum-safe cryptographic algorithms in TLS 1.3. This is achieved by building a Docker image that incorporates the OQS provider into the OpenSSL library used by Locust.

### Key Changes:

- Dockerfile Creation: A Dockerfile is included to build a custom Locust image using OpenSSL v3 with OQS support, allowing quantum-safe operations.
- Quantum-Safe Load Testing: Locust is now capable of negotiating quantum-safe keys and using quantum-safe authentication methods.
- Configuration Options: Several environment variables are introduced to control key settings such as log levels, the number of workers, the target host, and the cryptographic curves used (e.g., kyber768).

An example:
![Screenshot from 2024-10-17 17-20-06](https://github.com/user-attachments/assets/8f528369-7d50-4f09-b23e-dbb2cd2699d4)

davidgca commented:
> Thanks for the PR, @davidgca! Would you be willing to list yourself as a maintainer of this new demo in the README? Also, please feel free to add yourself to the list of contributors in the same file :)

I assume you say adding directly to README for OQS Demo? I also add a new line for the new tool in the table

davidgca commented on locust/scenarios/locustfile.py:
```diff
@@ -0,0 +1,88 @@
+                      f"User-Agent: {http_headers}\r\n"
+                      f"Connection: close\r\n\r\n",
+                capture_output=True, text=True
+            )
```
As I comment briefly in the README, I decided to use subprocess with openssl s_client in this implementation to handle post-quantum cryptographic curves. The reason for this approach is that current Python libraries, such as urllib and gevent, do not yet support these cryptographic standards.

By invoking openssl through subprocess, we can leverage its built-in support for post-quantum curves and ensure secure communication during our performance tests. This allows us to test the behavior of our endpoints with these  security protocols, without waiting for native support in Python libraries.

I also performed a test by recompiling Python with the OQS (Open Quantum Safe) version of OpenSSL to enable the necessary post-quantum curves. However, even as Python libraries evolve to support these curves, they will still rely on an underlying version of OpenSSL that supports post-quantum cryptography. So, even when native Python support is available, it will still be essential to ensure that the system's OpenSSL version has the required cryptographic capabilities.

baentsch replied:
Worthwhile taking a look at (maybe eventually using) https://github.com/open-quantum-safe/liboqs-python/pull/94?

davidgca replied:
Yes, I have looked into it, but after reviewing it, I believe this implementation requires a thorough investigation to assess whether this wrapper meets our performance standards (it's essential for any performance tool, to exhibit excellent performance behavior). For this initial version, I’d prefer not to include a wrapper without this research. Perhaps we can revisit this in future releases?

SWilson4 commented on locust/Dockerfile:
```diff
@@ -0,0 +1,111 @@
+# Take in all global args
+ARG LIBOQS_TAG
+ARG OQSPROVIDER_TAG
+ARG CURL_VERSION
```
Looks like this arg is never used—I'm assuming it's a copy-paste holdover from the curl dockerfile?

davidgca replied:
Yes it was a mistake during the first copy/paste, many thanks for all comments!!. 
Fix in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/91769ca63c25a4b9fb393634cf6ae2397f87d132)

SWilson4 commented on locust/Dockerfile:
```diff
@@ -0,0 +1,111 @@
+ARG PYTHON_VERSION
+ARG OPENSSL_BRANCH
+
+LABEL version="4"
```
This should probably be a lower number given that this is a new file, no?

davidgca replied:
Yes, set to 1 in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/91769ca63c25a4b9fb393634cf6ae2397f87d132)

SWilson4 commented on README.md:
```diff
@@ -17,23 +17,24 @@ We are explicitly soliciting contributors to maintain those integrations labelle
+| **Mosquitto**     | [Github: oqs-demos/mosquitto](mosquitto)                 | [Dockerhub: openquantumsafe/mosquitto](https://hub.docker.com/repository/docker/openquantumsafe/mosquitto)                                                                                                                                  | unsupported
+| **Envoy**         | [Github: oqs-demos/envoy](envoy)                         | [ Dockerhub: openquantumsafe/envoy](https://hub.docker.com/repository/docker/openquantumsafe/envoy)                                                                                                                                         | unsupported
+| **Unbound**       | [Github: oqs-demos/unbound](unbound)                     | [ Dockerhub: openquantumsafe/unbound](https://hub.docker.com/repository/docker/openquantumsafe/unbound)                                                                                                                                     | unsupported
+| **Locust**        | [Github: oqs-demos/locust](locust)                       | [ Dockerhub: openquantumsafe/locust](https://hub.docker.com/repository/docker/openquantumsafe/locust)                                                                                                                                       | @davidgca
```
Given that this demo is currently working and supported, I think it can move closer to the top of the list :)

davidgca replied:
Fix in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/91769ca63c25a4b9fb393634cf6ae2397f87d132)

SWilson4 commented on locust/README.md:
```diff
@@ -0,0 +1,41 @@
+
+1) Be sure to have [docker installed](https://docs.docker.com/install). 
+2) Run `docker build -t oqs-locust:0.0.1 .` to create a post quantum-enabled Locust docker image.
+3) In order to configure endpoints and their weight, modify the file [scerarios/locusfile.py](scenarios/locustfile.py), more information can be found in [USAGE.md](USAGE.md)
```
```suggestion
3) In order to configure endpoints and their weight, modify the file [scenarios/locusfile.py](scenarios/locustfile.py), more information can be found in [USAGE.md](USAGE.md)
```

davidgca replied:
Fix in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/91769ca63c25a4b9fb393634cf6ae2397f87d132)

SWilson4 commented on locust/README.md:
````diff
@@ -0,0 +1,41 @@
+3) In order to configure endpoints and their weight, modify the file [scerarios/locusfile.py](scenarios/locustfile.py), more information can be found in [USAGE.md](USAGE.md)
+4) To verify all components perform quantum-safe operations, first start the container with docker compose 
+
+```LOGGER_LEVEL=DEBUG HOST=https://qsc-nginx.discovery.hi.inet:4433 docker compose  up --scale master=1 --scale worker=8```
````
I think adding newlines before and after the command (inside the ```) tags will help with the display.

davidgca replied:
I agree, fix in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/91769ca63c25a4b9fb393634cf6ae2397f87d132)

SWilson4 commented on locust/USAGE.md:
There is quite a bit of duplication between this file and locust/README.md. Could they be restructured to avoid this?

davidgca replied:
Done in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/91769ca63c25a4b9fb393634cf6ae2397f87d132)

SWilson4 commented on locust/USAGE.md:
```diff
@@ -0,0 +1,55 @@
+Assuming Docker is [installed](https://docs.docker.com/install) the following command
+
+2) Run `docker build -t oqs-locust:0.0.1 .` to create a post quantum-enabled Locust docker image.
+3) To verify all components perform quantum-safe operations, first start the container with docker compose `LOGGER_LEVEL=DEBUG HOST=https://qsc-nginx.discovery.hi.inet:4433 docker compose  up --scale master=1 --scale worker=8`.
```
Where does this host URL come from?

davidgca replied:
It is my internal OQS host URL, Fix in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/91769ca63c25a4b9fb393634cf6ae2397f87d132)

SWilson4 reviewed (commented):
Hi @davidgca, thanks for the updates. I left a few style-related comments/questions.

I also followed the instructions in locust/README.md but got a 100% error rate. Not sure if this is expected?

davidgca commented:
> Thanks for the PR, @davidgca! Would you be willing to list yourself as a maintainer of this new demo in the README? Also, please feel free to add yourself to the list of contributors in the same file :)

davidgca commented:
> Hi @davidgca, thanks for the updates. I left a few style-related comments/questions.
> 
> I also followed the instructions in locust/README.md but got a 100% error rate. Not sure if this is expected?

Not at all, 100% error not sounds good!. The first step (and most important) is build the docker image:
`docker build -t oqs-locust:0.0.1 . `
what error do you have? I have recompiled this image(using my linux laptop) and I don't have any issue, maybe something with the docker version (mine is 27.3.1), maybe some stuff with docker and ARM cores(if  you are using an apple laptop )?

davidgca commented:
Hi @SWilson4 , I have signed-off all commits, do we need something else? Have you check if all Readme steps is working?

ajbozarth commented on locust/Dockerfile:
```diff
@@ -0,0 +1,110 @@
+# Define the degree of parallelism when building the image; leave the number away only if you know what you are doing
+ARG MAKE_DEFINES="-j 16"
+
+FROM alpine:3.20.3
```
```suggestion
FROM alpine:${ALPINE_VERSION} 
```

davidgca replied:
Added in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/042df20393d0704e6473d9ac5f87dd7c3856aeb2)

ajbozarth reviewed (commented):
I'm currently working on moving all the demos to pinning on the latest version (and double checking arm/amd support) in #298 It would help a bunch if you could make the same changes here (otherwise I'll just make the updates once this is merged). To help I noted them individually below:

davidgca commented:
> I'm currently working on moving all the demos to pinning on the latest version (and double checking arm/amd support) in #298 It would help a bunch if you could make the same changes here (otherwise I'll just make the updates once this is merged). To help I noted them individually below:

Thanks  for take a look on this. I will take all this suggestions in a single commit

davidgca commented on locust/Dockerfile:
```diff
@@ -0,0 +1,110 @@
+ARG LIBOQS_TAG=main
+
+# define the oqsprovider tag to be used
+ARG OQSPROVIDER_TAG=main
```
Added in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/042df20393d0704e6473d9ac5f87dd7c3856aeb2)

davidgca commented on locust/Dockerfile:
```diff
@@ -0,0 +1,110 @@
+# build OpenSSL3
+WORKDIR /opt/openssl
+RUN LDFLAGS="-Wl,-rpath -Wl,${INSTALLDIR}/lib64" ./config shared enable-zlib no-comp --prefix=${INSTALLDIR} && \
+    make ${MAKE_DEFINES} && make install_sw install_ssldirs install_dev;
```
Added in this [commit](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/042df20393d0704e6473d9ac5f87dd7c3856aeb2), tested with my amd64, please @SWilson4, assuming you are using a mac,  could you please verify if now the build is working

baentsch replied:
One more reason to suggest/request a (CI) test (build) (on x64 and aarch64)... ToDo for next PR, I understand @davidgca . @SWilson4 or @ajbozarth please resolve conversation if this works for you so we can move towards final PR approval.

davidgca commented on locust/Dockerfile:
```diff
@@ -0,0 +1,110 @@
+ARG OQSPROVIDER_TAG=main
+
+# define the Curl version to be baked in
+ARG OPENSSL_BRANCH="master"
```
done

baentsch commented on README.md:
```diff
@@ -17,23 +17,25 @@ We are explicitly soliciting contributors to maintain those integrations labelle
+| **HAproxy**       | [Github: oqs-demos/haproxy](haproxy)                     | [Dockerhub: openquantumsafe/haproxy](https://hub.docker.com/repository/docker/openquantumsafe/haproxy)                                                                                                                                      | unsupported
+| **Mosquitto**     | [Github: oqs-demos/mosquitto](mosquitto)                 | [Dockerhub: openquantumsafe/mosquitto](https://hub.docker.com/repository/docker/openquantumsafe/mosquitto)                                                                                                                                  | unsupported
+| **Envoy**         | [Github: oqs-demos/envoy](envoy)                         | [ Dockerhub: openquantumsafe/envoy](https://hub.docker.com/repository/docker/openquantumsafe/envoy)                                                                                                                                         | unsupported
+| **Unbound**       | [Github: oqs-demos/unbound](unbound)                     | [ Dockerhub: openquantumsafe/unbound](https://hub.docker.com/repository/docker/openquantumsafe/unbound)                                                                                                                                     | unsupported
```
Why did all these lines change and not just one?

davidgca replied:
No idea, I 've just added a new line in the middle of the table, maybe github are doing any stuff comparing changes inside a table for a .md file?

baentsch replied:
In a word: Hmpf.

baentsch commented on locust/README.md:
```diff
@@ -0,0 +1,30 @@
+## Purpose 
+This directory contains a Dockerfile that builds `Locust` using OpenSSL v3 using the [OQS provider](https://github.com/open-quantum-safe/oqs-provider) and Python3, which allows `Locust` to negotiate quantum-safe keys and use quantum-safe authentication in TLS 1.3.
```
Suggest adding a link to the upstream project here

davidgca replied:
Done [here](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/ab1d2f45c378d14f7b5c01a1e6a3b656a69b8b3c#diff-8a1eed52bbc906e4f897ac6f113321abdd9f6fa9c742de72ac721a6d0cb6f195R4)

baentsch commented on locust/USAGE.md:
```diff
@@ -0,0 +1,56 @@
+- LOGGER_LEVEL: Set the log level for the locust master and worker. Default is ERROR.
+- HOST: Set the host to test. Default is https://test:4433
+- WORKERS: Set the number of workers. Default is 8. Ideally, the number of workers should be the same as the number of cores in the machine.
+- MASTER: Set the number of master. Default is 1.
```
What is the number of master? Any number possible? What's the rationale for that? Sorry for the possibly pointless question from a person not knowing `locust` -- but then others may have the same question. So a pointer to the documentation of thus number may be helpful.

davidgca replied:
Please feel free to ask any questions; I'm happy to provide answers. You’re absolutely right—this could indeed lead to errors. Locust, like any other master-worker cluster setup, allows only one master and multiple workers (in a previous project, I managed to run 800 workers across 200 VMs with obviously one unique master). I exposed the master variable unnecessarily, so I’ll modify it so that with docker-compose, we only need to specify the number of workers

davidgca replied:
Done [here](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/ab1d2f45c378d14f7b5c01a1e6a3b656a69b8b3c#diff-8a1eed52bbc906e4f897ac6f113321abdd9f6fa9c742de72ac721a6d0cb6f195R4)

baentsch commented on locust/scenarios/locustfile.py:
```diff
@@ -0,0 +1,88 @@
+logger = logging.getLogger(__name__)
+logger.setLevel(logger_level)
+
+curve = str(os.environ.get("CURVE"))
```
"Curve" is a pretty dated term from the time KEM could only do EC curves. Suggest renaming to "Group" which is the more standard term (also used in TLS1.3).

davidgca replied:
Done [here](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/ab1d2f45c378d14f7b5c01a1e6a3b656a69b8b3c#diff-8a1eed52bbc906e4f897ac6f113321abdd9f6fa9c742de72ac721a6d0cb6f195R4)

baentsch reviewed (changes requested):
Thanks for the integration @davidgca ! Please see my single comments. Please seriously consider changing the term curve to group. If there were interest to put this on docker hub (?) so users don't have to build this but can just "get going", please also add the corresponding CI build job.

davidgca commented on locust/README.md:
```diff
@@ -0,0 +1,30 @@
+
+1) Be sure to have [docker installed](https://docs.docker.com/install). 
+2) Run `docker build -t oqs-locust:0.0.1 .` to create a post quantum-enabled Locust docker image.
+3) In order to configure endpoints and their weight, modify the file [scenarios/locusfile.py](scenarios/locustfile.py), more information can be found in [USAGE.md](USAGE.md)
```
Done [here](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/ab1d2f45c378d14f7b5c01a1e6a3b656a69b8b3c#diff-8a1eed52bbc906e4f897ac6f113321abdd9f6fa9c742de72ac721a6d0cb6f195R4)

davidgca commented:
> Thanks for the integration @davidgca ! Please see my single comments. Please seriously consider changing the term curve to group. If there were interest to put this on docker hub (?) so users don't have to build this but can just "get going", please also add the corresponding CI build job.

As I mentioned before, I've already modified this and refactored it by changing 'curve' to 'group.' Thanks a lot for this comment! I actually had some doubts about considering Kyber a curve, as my confusion came from OpenSSL using the 'curve' option for any post-quantum scheme

baentsch commented on README.md:
```diff
@@ -17,23 +17,25 @@ We are explicitly soliciting contributors to maintain those integrations labelle
+| **Apache httpd**  | [Github: oqs-demos/httpd](httpd)                         | [Dockerhub: openquantumsafe/httpd](https://hub.docker.com/repository/docker/openquantumsafe/httpd)                                                                                                                                          | @baentsch
+| **nginx**         | [Github: oqs-demos/nginx](nginx)                         | [Dockerhub: openquantumsafe/nginx](https://hub.docker.com/repository/docker/openquantumsafe/nginx), [Dockerhub: openquantumsafe/nginx-quic](https://hub.docker.com/repository/docker/openquantumsafe/nginx-quic)                            | @baentsch, @bhess, @pi-314159
+| **Chromium**      | [Github: oqs-demos/chromium](chromium) (limited support) | -                                                                                                                                                                                                                                           | @pi-314159 |
+| **Locust**        | [Github: oqs-demos/locust](locust)                       | [ Dockerhub: openquantumsafe/locust](https://hub.docker.com/repository/docker/openquantumsafe/locust)                                                                                                                                       | @davidgca|
```
Now reviewing all these lines one by one manually -- and that answers another of my questions: This _is_ targeted for publication on docker hub (which IMO is sensible) -- but then this also requires a (test-)build-and-push instruction in CI. Please add that, e.g., to [linux.yml](https://github.com/open-quantum-safe/oqs-demos/blob/main/.github/workflows/linux.yml)

davidgca replied:
Sorry I'm not familiar to github workflow, and how can I test an eventually new CI,  this is mandatory? could we add docker image publication for next PR?

baentsch replied:
Sure. @ajbozarth anyway volunteered to redo CI in this space afaik. But in that case, please remove the reference to the docker image at docker hub in this PR.

davidgca replied:
remove it [here](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/ad093355501e576ec1eeffdf9e015781b0eb485e)
Thanks @baentsch  I want to complete this PR and this could maybe extend this PR unnecessary. I will take a look about your CI steps after merge this PR

baentsch commented:
Thanks for these updates, @davidgca !

> as my confusion came from OpenSSL using the 'curve' option for any post-quantum scheme

That should not be the case any more... Can I ask where we do that? Direct links would be best such as to know where to change. Thanks in advance!

davidgca commented:
> Thanks for these updates, @davidgca !
> 
> > as my confusion came from OpenSSL using the 'curve' option for any post-quantum scheme
> 
> That should not be the case any more... Can I ask where we do that? Direct links would be best such as to know where to change. Thanks in advance!

sure, [here](https://github.com/open-quantum-safe/oqs-demos/blob/ab1d2f45c378d14f7b5c01a1e6a3b656a69b8b3c/locust/scenarios/locustfile.py#L37), As you cans see, we are using `openssl s_client -curves`

baentsch commented:
> sure, [here](https://github.com/open-quantum-safe/oqs-demos/blob/ab1d2f45c378d14f7b5c01a1e6a3b656a69b8b3c/locust/scenarios/locustfile.py#L37), As you cans see, we are using `openssl s_client -curves`

Well, that's your choice, isn't it? In OpenSSL this command line option is kept for backwards compatibility with TLS1.2; for TLS1.3 (which is the only spec supporting PQC) the term (and option to use) is "-groups" (https://docs.openssl.org/master/man3/SSL_CONF_cmd/#supported-command-line-commands btw lists "-curves" only as a synonym ... but looking at the documentation I think it would be prudent to add verbiage that providers can extend the list given there; thus, thanks for giving me reason to check @davidgca).

davidgca commented:
> > sure, [here](https://github.com/open-quantum-safe/oqs-demos/blob/ab1d2f45c378d14f7b5c01a1e6a3b656a69b8b3c/locust/scenarios/locustfile.py#L37), As you cans see, we are using `openssl s_client -curves`
> 
> Well, that's your choice, isn't it? In OpenSSL this command line option is kept for backwards compatibility with TLS1.2; for TLS1.3 (which is the only spec supporting PQC) the term (and option to use) is "-groups" (https://docs.openssl.org/master/man3/SSL_CONF_cmd/#supported-command-line-commands btw lists "-curves" only as a synonym ... but looking at the documentation I think it would be prudent to add verbiage that providers can extend the list given there; thus, thanks for giving me reason to check @davidgca).

Thanks! In my initial research, I saw the curves option and didn't look into the groups option. I've tested it successfully and made the change [here](https://github.com/open-quantum-safe/oqs-demos/pull/304/commits/f35b0fa362ed8c6c14e8c57dc0bf63206a9efb19)

baentsch commented:
> Hi @SWilson4 and @ajbozarth, could we go ahead and merge this PR? Or is there anything you think might still be missing?"

Ping @SWilson4 @ajbozarth : Objections to merge?

davidgca commented:
Many thanks for all these comments and suggestions, and your approvals. I see on this [action](https://github.com/open-quantum-safe/oqs-demos/actions/runs/11575837654/job/32302453547?pr=304) that workflow is failing with:
```Run docker/login-action@v2
Error: Username and password required```
Anything I can do? @baentsch  could you help to fix this?

SWilson4 commented:
> Many thanks for all these comments and suggestions, and your approvals. I see on this [action](https://github.com/open-quantum-safe/oqs-demos/actions/runs/11575837654/job/32302453547?pr=304) that workflow is failing with:
> 
> ```
> Error: Username and password required```
> Anything I can do? @baentsch  could you help to fix this?
> ```

That's to be expected; our CI run unfortunately fails unless you have sufficient perms to access repository secrets. I believe @ajbozarth will be looking at a fix for this in the near future.

ajbozarth commented:
> I believe @ajbozarth will be looking at a fix for this in the near future.

That is correct, it is one of the pain points I most was to address in my upcoming CI work.

> our CI run unfortunately fails unless you have sufficient perms to access repository secrets

IIUC I believe you can input the expected secrets in GitHub Actions for your own fork and it should work if you really want to

baentsch commented:
Merged. Thanks for the contribution @davidgca !
